node/actors/network/src/metrics.rs

```diff
@@ -245,58 +253,58 @@ impl MeteredStreamStats {
             received: 0.into(),
             established: SystemTime::now(),
             peer_addr,
-            sent_buffer: VecDeque::new(),
-            received_buffer: VecDeque::new(),
+            current_minute_sent: 0.into(),
+            previous_minute_sent: 0.into(),
+            current_minute_received: 0.into(),
+            previous_minute_received: 0.into(),
+            current_minute_start: 0.into(),
         }
     }
 
-    fn read(&mut self, amount: u64) {
+    fn read(&self, amount: u64) {
+        self.update_minute();
         self.received.fetch_add(amount, Ordering::Relaxed);
-        self.received_buffer.push_back((Instant::now(), amount));
-        self.cleanup_buffers();
+        self.current_minute_received
+            .fetch_add(amount, Ordering::Relaxed);
     }
 
-    fn wrote(&mut self, amount: u64) {
+    fn wrote(&self, amount: u64) {
+        self.update_minute();
         self.sent.fetch_add(amount, Ordering::Relaxed);
-        self.sent_buffer.push_back((Instant::now(), amount));
-        self.cleanup_buffers();
+        self.current_minute_sent
+            .fetch_add(amount, Ordering::Relaxed);
     }
 
-    fn cleanup_buffers(&mut self) {
-        let one_minute_ago = Instant::now() - Duration::from_secs(60);
+    fn update_minute(&self) {
+        let now = SystemTime::now()
+            .duration_since(UNIX_EPOCH)
+            .expect("Time went backwards")
+            .as_secs();
 
-        while let Some(&(timestamp, _)) = self.sent_buffer.front() {
-            if timestamp < one_minute_ago {
-                self.sent_buffer.pop_front();
-            } else {
-                break;
-            }
-        }
+        let elapsed_secs = now.saturating_sub(self.current_minute_start.load(Ordering::Relaxed));
 
-        while let Some(&(timestamp, _)) = self.received_buffer.front() {
-            if timestamp < one_minute_ago {
-                self.received_buffer.pop_front();
-            } else {
-                break;
-            }
+        if elapsed_secs >= 60 {
+            self.previous_minute_sent.store(
+                self.current_minute_sent.load(Ordering::Relaxed),
+                Ordering::Relaxed,
+            );
+            self.previous_minute_received.store(
+                self.current_minute_received.load(Ordering::Relaxed),
+                Ordering::Relaxed,
+            );
+            self.current_minute_sent.store(0, Ordering::Relaxed);
+            self.current_minute_received.store(0, Ordering::Relaxed);
+            self.current_minute_start.store(now, Ordering::Relaxed);
         }
     }
 
     /// Returns the upload throughput of the connection in bytes per second.
     pub fn sent_throughput(&self) -> f64 {
-        self.sent_buffer
-            .iter()
-            .map(|&(_, amount)| amount)
-            .sum::<u64>() as f64
-            / 60.0
+        self.previous_minute_sent.load(Ordering::Relaxed) as f64 / 60.0
```

Comment:
note that if no data is transferred for a minute, then this value will not reset to 0. You should compare the current timestamp against the current_minute_start and based on that return either previous_minute_sent/current_minute_sent or 0.

Reply:
Or I can just call update_minute() in the beginning.

Reply:
that might cause race condition when running update_minute() minute from read()/write() and http rendering function. If you want to do that, then put a mutex instead of bunch of atomic vars.